pallets/bonds/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+		type TimestampProvider: Time<Moment = Moment>;
+
+		/// The pallet id, used for deriving its sovereign account ID.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Min time for maturity.
+		#[pallet::constant]
+		type MinMaturity: Get<Moment>;
+
+		/// The origin which can issue new bonds.
+		type IssueOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+
+		/// The origin which can unlock bonds by making them mature.
+		type UnlockOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Protocol fee.
+		#[pallet::constant]
+		type ProtocolFee: Get<Permill>;
+
+		/// Protocol fee receiver.
+		#[pallet::constant]
+		type FeeReceiver: Get<Self::AccountId>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::storage]
+	/// Registered bonds. Fully redeemed bonds are removed from the storage.
+	#[pallet::getter(fn bonds)]
+	pub(super) type Bonds<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, Bond<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A bond asset was registered
+		BondTokenCreated {
+			issuer: T::AccountId,
+			asset_id: T::AssetId,
+			bond_asset_id: T::AssetId,
+			amount: T::Balance,
+			fee: T::Balance,
+		},
+		/// Bonds were redeemed
+		BondsRedeemed {
+			who: T::AccountId,
+			bond_id: T::AssetId,
+			amount: T::Balance,
+		},
+		/// Bonds were unlocked
+		BondsUnlocked { bond_id: T::AssetId },
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Bond not registered
+		BondNotRegistered,
+		/// Bond is not mature
+		BondNotMature,
+		/// Maturity not long enough
+		InvalidMaturity,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Issue new bonds.
+		/// New asset id is registered and assigned to the bonds.
+		/// The number of bonds the issuer receives is 1:1 to the `amount` of the underlying asset
+		/// minus the protocol fee.
+		/// The bond asset is registered with the empty string for the asset name,
+		/// and with the same existential deposit as of the underlying asset.
+		/// Bonds can be redeemed 1:1 for the underlying asset once mature.
+		/// Protocol fee is applied to the amount, and transferred to `T::FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: issuer of new bonds, needs to be `T::IssueOrigin`
+		/// - `asset_id`: underlying asset id
+		/// - `amount`: the amount of the underlying asset
+		/// - `maturity`: Unix time in milliseconds, when the bonds will be mature. Needs to be set
+		/// more than `T::MinMaturity` from now.
+		///
+		/// Emits `BondTokenCreated` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::issue())]
+		pub fn issue(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			amount: T::Balance,
+			maturity: Moment,
+		) -> DispatchResult {
+			let who = T::IssueOrigin::ensure_origin(origin)?;
```

Comment:
anyone can issue bonds

Reply:
ok. I will remove the config param. The idea was to have it configurable, but configure it so that everyone can do that.

Reply:
but we could keep it if we plan to enable this for others only later